### PR TITLE
Add Update Factor Admin endpoint

### PR DESCRIPTION
**Pull request by J0**
## What kind of change does this PR introduce?

**Endpoints** 
- [ ] UPDATE factor so status can be changed

## What is the current behavior?
No admin Endpoints

## What is the new behavior?

To be filled in

## Additional context

Relevant issue #14 

Still need an explicit /list factor

### CODE REVIEW

File: api/admin.go [hunk 1]
```diff
@@ -342,3 +351,80 @@ func (a *API) adminUserDelete(w http.ResponseWriter, r *http.Request) error {
...
+func (a *API) adminUserUpdateFactorStatus(w http.ResponseWriter, r *http.Request) error {
+	ctx := r.Context()
+	user := getUser(ctx)
+	instanceID := getInstanceID(ctx)
+
+	params := &AdminUserUpdateFactorStatusParams{}
+	jsonDecoder := json.NewDecoder(r.Body)
+	err := jsonDecoder.Decode(params)
+	if err != nil {
+		return badRequestError("Could not read AdminUpdateFactorStatus params: %v", err)
```
Comment (hf) [inline api/admin.go:364]: Since this message goes out to users, maybe disclose only the information they need to fix it?

File: api/admin.go [hunk 2]
```diff
@@ -342,3 +351,80 @@ func (a *API) adminUserDelete(w http.ResponseWriter, r *http.Request) error {
...
+func (a *API) adminUserDeleteFactor(w http.ResponseWriter, r *http.Request) error {
+	ctx := r.Context()
+	user := getUser(ctx)
+	instanceID := getInstanceID(ctx)
+
+	params := &AdminUserDeleteFactorParams{}
+	jsonDecoder := json.NewDecoder(r.Body)
+	err := jsonDecoder.Decode(params)
+	if err != nil {
+		return badRequestError("Could not read AdminUserDeleteFactor params: %v", err)
```
Comment (hf) [inline api/admin.go:405]: Ditto on error message for API users.

File: api/api.go [hunk 3]
```diff
@@ -169,6 +169,12 @@ func NewAPIWithVersion(ctx context.Context, globalConfig *conf.GlobalConfigurati
 					r.Get("/", api.adminUserGet)
 					r.Put("/", api.adminUserUpdate)
 					r.Delete("/", api.adminUserDelete)
+					r.Route("/mfa", func(r *router) {
+						r.Put("/update_factor_status", api.adminUserUpdateFactorStatus)
+						r.Delete("/delete_factor", api.adminUserDeleteFactor)
```
Comment (hf) [inline api/api.go:174]: Why not just use HTTP verbs on `/mfa`? Example:

```suggestion
						r.Put("/", api.adminUserUpdateFactorStatus)
						r.Delete("/", api.adminUserDeleteFactor)
```

File: api/api.go [hunk 4]
```diff
@@ -169,6 +169,12 @@ func NewAPIWithVersion(ctx context.Context, globalConfig *conf.GlobalConfigurati
...
+						r.Put("/disable", api.DisableMFA)
+						r.Put("/enable", api.EnableMFA)
```
Comment (hf) [inline api/api.go:176]: Same here -- can we make this more RESTful? Example: to disable a factor you send a PUT on `/mfa/<factor>`; or to delete it a DELETE on `/mfa/<factor>`?

Sorry if this is handled in the RFC, but it's kind-of standing out from the rest of the API in GoTrue.

File: api/mfa.go [hunk 5]
```diff
@@ -101,3 +127,72 @@ func (a *API) GenerateRecoveryCodes(w http.ResponseWriter, r *http.Request) erro
...
+	instanceID := getInstanceID(ctx)
+	if !user.MFAEnabled {
+		return MFANotEnabledError
+	}
+
+	params := &EnrollFactorParams{}
+	jsonDecoder := json.NewDecoder(r.Body)
+	err := jsonDecoder.Decode(params)
+	if err != nil {
+		return badRequestError("Could not read EnrollFactor params: %v", err)
```
Comment (hf) [inline api/mfa.go:145]: Ditto on error message for API users.

File: api/mfa.go [hunk 6]
```diff
@@ -101,3 +127,72 @@ func (a *API) GenerateRecoveryCodes(w http.ResponseWriter, r *http.Request) erro
...
+	}
+
+	if (params.FactorType != "totp") && (params.FactorType != "webauthn") {
+		return unprocessableEntityError("FactorType needs to be either 'totp' or 'webauthn'")
+	}
```
Comment (hf) [inline api/mfa.go:150]: Ditto on error message for API users and a switch statement.

File: api/api.go (hunk 3 above)
Comment (J0) [reply api/api.go]: challenged to "/" will do the same with the other  endpoints

**Comment by J0** [comment]
Closing since we overhauled the API routes  -- see #586 for update factor admin endpoint

**Pull request closed without merging**